internal/worker/authenticationworker/worker.go

```diff
@@ -119,24 +122,53 @@ func (kw *keyupdaterWorker) Handle(_ <-chan struct{}) error {
 		return err
 	}
 	// Figure out if any keys have been added or deleted.
-	newJujuKeys := set.NewStrings(newKeys...)
+	newJujuKeys := set.NewStrings(ensureJujuCommentForKeys(newKeys)...)
 	deleted := kw.jujuKeys.Difference(newJujuKeys)
 	added := newJujuKeys.Difference(kw.jujuKeys)
-	if added.Size() > 0 || deleted.Size() > 0 {
-		logger.Infof("adding ssh keys to authorised keys: %v", added)
-		logger.Infof("deleting ssh keys from authorised keys: %v", deleted)
-		if err = kw.writeSSHKeys(newKeys); err != nil {
-			err = errors.Annotate(err, "updating ssh keys")
-			logger.Infof(err.Error())
-			return err
+	err = ssh.AddKeys(SSHUser, added.Values()...)
+	if err != nil {
+		err = errors.Annotatef(err, "adding ssh keys to authorised keys: %v", added)
+	}
+	keyFingerprintsToDelete := []string{}
+	for _, key := range deleted.Values() {
+		fingerprint, _, err := ssh.KeyFingerprint(key)
+		if err != nil {
+			continue
 		}
+		keyFingerprintsToDelete = append(keyFingerprintsToDelete, fingerprint)
+	}
+
+	err = ssh.DeleteKeys(SSHUser, keyFingerprintsToDelete...)
+	if err != nil {
+		err = errors.Annotatef(err, "removing ssh keys from authorised keys: %v", deleted)
 	}
 	kw.jujuKeys = newJujuKeys
 	return nil
 }
 
 // TearDown is defined on the worker.NotifyWatchHandler interface.
-func (kw *keyupdaterWorker) TearDown() error {
+func (kw *keyUpdater) TearDown() error {
 	// Nothing to do here.
 	return nil
 }
+
+func (a *AuthWorker) AddEphemeralKey(ephemeralKey string) error {
```

Comment:
We need to check the worker isn't dead.

Reply:
Yep, that's fine then. Since we're consuming this from another worker, that depends on this, as long as it dies correctly, we don't need it.